pkg/updater/jobparser.go

```diff
@@ -0,0 +1,335 @@
+	ports := make([]corev1.ContainerPort, 0)
+	basePort := int32(job.Spec.Port)
+	for i := 0; i < portsTotal; i++ {
+		ports = append(ports, corev1.ContainerPort{
+			Name:          fmt.Sprintf("jobport-%d", basePort),
+			ContainerPort: basePort,
+		})
+		basePort++
+	}
+	return ports
+}
+
+func masterPorts(job *paddlev1.TrainingJob) []corev1.ContainerPort {
+	ports := []corev1.ContainerPort{
+		corev1.ContainerPort{
+			Name:          "master-port",
+			ContainerPort: 8080,
+		},
+		corev1.ContainerPort{
+			Name:          "etcd-port",
+			ContainerPort: 2379,
+		},
+	}
+	return ports
+}
+
+func podEnv(job *paddlev1.TrainingJob) []corev1.EnvVar {
+	needGPU := "0"
+	if job.NeedGPU() {
+		needGPU = "1"
+	}
+	trainerCount := 1
+	if job.NeedGPU() {
+		q := job.Spec.Trainer.Resources.Requests.NvidiaGPU()
+		trainerCount = int(q.Value())
+	} else {
+		q := job.Spec.Trainer.Resources.Requests.Cpu()
+		// FIXME: CPU resource value can be less than 1.
+		trainerCount = int(q.Value())
+	}
+
+	return []corev1.EnvVar{
+		corev1.EnvVar{Name: "PADDLE_JOB_NAME", Value: job.ObjectMeta.Name},
+		// NOTICE: TRAINERS, PSERVERS, PADDLE_INIT_NUM_GRADIENT_SERVERS
+		//         these env are used for non-faulttolerant training,
+		//         use min-instance all the time. When job is elastic,
+		//         these envs are not used.
+		corev1.EnvVar{Name: "TRAINERS", Value: strconv.Itoa(job.Spec.Trainer.MinInstance)},
+		corev1.EnvVar{Name: "PSERVERS", Value: strconv.Itoa(job.Spec.Pserver.MinInstance)},
+		corev1.EnvVar{Name: "ENTRY", Value: job.Spec.Trainer.Entrypoint},
+		// FIXME: TOPOLOGY deprecated
+		corev1.EnvVar{Name: "TOPOLOGY", Value: job.Spec.Trainer.Entrypoint},
+		corev1.EnvVar{Name: "TRAINER_PACKAGE", Value: job.Spec.Trainer.Workspace},
+		corev1.EnvVar{Name: "PADDLE_INIT_PORT", Value: strconv.Itoa(job.Spec.Port)},
+		// PADDLE_INIT_TRAINER_COUNT should be same to gpu number when use gpu
+		// and cpu cores when using cpu
+		corev1.EnvVar{Name: "PADDLE_INIT_TRAINER_COUNT", Value: strconv.Itoa(trainerCount)},
+		corev1.EnvVar{Name: "PADDLE_INIT_PORTS_NUM", Value: strconv.Itoa(job.Spec.PortsNum)},
+		corev1.EnvVar{Name: "PADDLE_INIT_PORTS_NUM_FOR_SPARSE", Value: strconv.Itoa(job.Spec.PortsNumForSparse)},
+		corev1.EnvVar{Name: "PADDLE_INIT_NUM_GRADIENT_SERVERS", Value: strconv.Itoa(job.Spec.Trainer.MinInstance)},
+		corev1.EnvVar{Name: "PADDLE_INIT_NUM_PASSES", Value: strconv.Itoa(job.Spec.Passes)},
+		corev1.EnvVar{Name: "PADDLE_INIT_USE_GPU", Value: needGPU},
+		corev1.EnvVar{Name: "LD_LIBRARY_PATH", Value: "/usr/local/cuda/lib64"},
+		corev1.EnvVar{Name: "NAMESPACE", ValueFrom: &corev1.EnvVarSource{
+			FieldRef: &corev1.ObjectFieldSelector{
+				FieldPath: "metadata.namespace",
+			},
+		}},
+		corev1.EnvVar{Name: "POD_IP", ValueFrom: &corev1.EnvVarSource{
+			FieldRef: &corev1.ObjectFieldSelector{
+				FieldPath: "status.podIP",
+			},
+		}},
+	}
+}
+
+// -----------------------------------------------------------------------
```

Comment:
Go's comment should follow Godoc's convention: https://blog.golang.org/godoc-documenting-go-code.